tensorflow/python/data/kernel_tests/iterator_test.py

```diff
@@ -1016,6 +1018,85 @@ def fn():
     self.evaluate(counter_var.initializer)
     self.assertEqual(self.evaluate(fn()), 10)
 
+  def assert_dataset_placement(self, host_dataset, host_iterator, host_tensor,
+                               device_dataset, device_iterator, device_tensor):
+
+    def assert_host_placement(_obj):
+      try:
+        self.assertIn("CPU:0", _obj)
+      except AssertionError:
+        self.assertEqual(_obj, "")
+
+    assert_host_placement(host_dataset._variant_tensor.device)
+    assert_host_placement(host_tensor.device)
+
+    self.assertIn("GPU:0", device_dataset._variant_tensor.device)
+    self.assertIn("GPU:0", device_tensor.device)
+
+    if not tf2.enabled() or context.executing_eagerly():
```

Comment:
If I do check check device I obtain the error that the attribute `_device` does not exist in this scenario. Sounds like an issue to me but I didn't want to fix something disconnect from this PR.

Reply:
This is somewhat worrisome because if the test code cannot access `.device` that neither will non-test code and since the code you have modified will run in TF2 graph mode, I suspect that this is an issue.

@alextp does it make sense that the `.device` attribute is no present in TF2 graph mode?

Reply:
The Tensor.device attribute is present in tf2 graph mode. Or are we not getting the device of a tensor? Can you print the object which doesn't have the device attribute on a failure and paste the result here or in a gist?